### PR TITLE
Add vTPM support for Linux

### PR DESCRIPTION
**Pull request by stefanberger**
This patch series adds vTPM support for Linux to libcontainer.

The functionality is based on a recently added vtpm_proxy driver, which is becoming
available in Linux 4.8. The driver provides /dev/vtpmx, on which an ioctl is called
that spawns a TPM device in the host's /dev directory and returns an anonymous file-
descriptor on which a TPM emulator can listen for TPM commands. If we for example
created /dev/tpm12 on the host we make this device available as /dev/tpm0 inside the
container. We also add its major and minor numbers to the device cgroup.

We implement a VTPM class that allows us to create the device and starts a TPM
emulator 'swtpm', to which it passes the anonymous file descriptor.
Besides that, the user can choose to have the vTPM create certificates in a step
that simulates TPM manufacturing. We do this by calling the external swtpm_setup
program, which is part of the swtpm project.

VTPM support is added inside the JSON configuration as follows:

```
[...]
    "linux": {
        "resources": {
            "devices": [
                {
                    "allow": false,
                    "access": "rwm"
                }
            ] ,
            "vtpms": [
                {
                    "statePath": "/tmp/tpm-1",
                    "createCerts": true,
                },
            ]
        },
```

This JSON markup makes a single TPM available inside the created container.

o The statPath parameter indicates the directory where the TPM emulator 'swtpm'
  writes the state of the TPM device to.
o The createCerts parameter indicates that certificates for the TPM are
  to be created.

The swtpm project is available here  : https://github.com/stefanberger/swtpm
The libtpms project is available here: https://github.com/stefanberger/libtpms

Signed-off-by: Stefan Berger stefanb@linux.vnet.ibm.com

### CODE REVIEW

**Comment by stefanberger** [comment]
This is a continuation of the PR https://github.com/opencontainers/runc/pull/1591 that cannot be reopened anymore.

**Comment by deitch** [comment]
Thanks for reopening.

File: vendor/github.com/opencontainers/runtime-spec/specs-go/config.go [hunk 1]
```diff
@@ -320,6 +320,13 @@ type LinuxNetwork struct {
 	Priorities []LinuxInterfacePriority `json:"priorities,omitempty"`
 }
 
+// VTPM definition
+type VTPM struct {
```
Comment (cyphar) [inline vendor/github.com/opencontainers/runtime-spec/specs-go/config.go]: @crosbymichael How do you feel about including changes to `vendor/`?
@stefanberger Is there a runtime-spec PR open about this?

Comment (stefanberger) [reply vendor/github.com/opencontainers/runtime-spec/specs-go/config.go]: @cyphar I opened https://github.com/opencontainers/runtime-spec/pull/920 for vTPM support

**Comment by lukehinds** [comment]
are you working on this still @stefanberger ? I don't have reviewer access, but from what I have looked at, the patch looks in good shape.

**Comment by JAORMX** [comment]
Seems there are merge conflicts here. Anybody working on this still?

**Comment by stefanberger** [comment]
I rebased this on the latest master and am now creating the vTPM devices at higher layer in `utils_linux.go`, which makes cleanup in the error case less invasive.

**Comment by stefanberger** [comment]
This now works right out of the box with a recent kernel on Fedora 31. The kernel-modules-extra package for `5.3.16-300.fc31` has the necessary kernel module. 

Install the packages:
`dnf -y install swtpm-tools`

Create a spec:
`/usr/loca/sbinl/runc spec`

Add the following modifications to the spec in `config.json`:
```
                                        "allow": false,
                                        "access": "rwm"
                                }
+                       ],
+                       "vtpms": [
+                               {
+                                       "statepath": "/tmp/myvtpm1",
+                                       "vtpmversion": "2"
+                               }
                        ]
                },
                "namespaces": [
```

Run a container:
```
/usr/local/sbin/runc run test
sh-5.0# ls -l /dev/tpm*
crw-------. 1 root root 253,     1 Jan  2 14:59 /dev/tpm0
crw-------. 1 root root 253, 65537 Jan  2 14:59 /dev/tpmrm0
```

**Comment by stefanberger** [comment]
I just extended the series with patches for running the TPM emulator `swtpm` with an AppArmor profilevcreated on the fly. I had tried to use the existing function `apparmor.ApplyProfile()` that tries to apply a profile via writing to `/proc/self/attr/exec`, though this created 50% failures. I have to write to `/proc/<tid>/attr/exec` to get 100% success.

**Comment by stefanberger** [comment]
Running `swtpm` with an SELinux label has now also landed:

```
# ps auxZ | grep swtpm
system_u:system_r:container_t:s0:c494,c987 root 1149416 1.2  0.0 7680 3104 ?     Ss   09:20   0:00 swtpm chardev --tpmstate dir=/var/lib/runc/myvtpm1 --daemon --fd 5 --pid file=/var/lib/runc/myvtpm1/swtpm.pid --log file=/var/lib/runc/myvtpm1/swtpm.log --runas root --tpm2 --locality reject-locality-4,allow-set-locality --flags not-need-init
# sudo ls -lZ /var/lib/runc/myvtpm
total 20
drwxr-x---. 2 root root system_u:object_r:container_file_t:s0:c927,c931 4096 Jan  6 09:21 .
drwxr-xr-x. 3 root root unconfined_u:object_r:var_lib_t:s0              4096 Jan  4 19:46 ..
-rw-r-----. 1 root root system_u:object_r:container_file_t:s0:c927,c931    0 Jan  6 09:21 .lock
-rw-r--r--. 1 root root system_u:object_r:container_file_t:s0:c927,c931   75 Jan  4 19:46 swtpm.log
-rw-r--r--. 1 root root system_u:object_r:container_file_t:s0:c927,c931    7 Jan  6 09:21 swtpm.pid
-rw-r-----. 1 root root system_u:object_r:container_file_t:s0:c927,c931 1187 Jan  6 09:21 tpm2-00.permall

```

**Comment by stefanberger** [comment]
... now also supports vTPM state encryption

**Comment by stefanberger** [comment]
@vbatts I think the vTPM support is now quite complete. There are patches that would need to go into the runtime-spec that are currently only here, but I will move them over only once someone tells me this is going to go in.

**Comment by vbatts** [comment]
On 08/01/20 08:22 -0800, Stefan Berger wrote:
>   [1]@vbatts I think the vTPM support is now quite complete. There are
>   patches that would need to go into the runtime-spec that are currently
>   only here, but I will move them over only once someone tells me this is
>   going to go in.

I'm not against it. The conversation just didn't come to a conclusion.

**Comment by stefanberger** [comment]
> I'm not against it. The conversation just didn't come to a conclusion.

@vbatts Who can you/we ask?

**Comment by stefanberger** [comment]
@cyphar Do you have comments on this series?

**Comment by vbatts** [comment]
@opencontainers/runc-maintainers please review this. I've now heard from a couple sources to get eyes on this feature.

**Comment by vbatts** [comment]
(also kick the CI. Looks like the VM is stuck booting?)

**Comment by cyphar** [comment]
> (also kick the CI. Looks like the VM is stuck booting?)

It looks like that test is super flaky (it's been failing for basically every PR until you restart it a few times). I'll send a PR to add it to the list of allowable failures.

**Comment by lukehinds** [comment]
How are we doing for this PR now and what's our main factors we need to address to get this merged..do tests need more work?

**Comment by stefanberger** [comment]
To be upfront: running it as container in container doesn't work. The top level container would have to create the vTPM instances and then pass those through to its children as a chardev. Though I haven't tried this myself.

File: libcontainer/vtpm/vtpm.go [hunk 2]
```diff
@@ -0,0 +1,893 @@
...
+	"time"
+	"unsafe"
+
+	"github.com/opencontainers/runc/libcontainer/apparmor"
+	selinux "github.com/opencontainers/selinux/go-selinux"
+
+	"github.com/sirupsen/logrus"
+)
+
+// object
```
Comment (AkihiroSuda) [inline libcontainer/vtpm/vtpm.go]: This godoc comment doesn't seem meaningful

File: libcontainer/vtpm/vtpm.go [hunk 3]
```diff
@@ -0,0 +1,893 @@
...
+type vtpm_proxy_new_dev struct {
+	flags       uint32
+	tpm_dev_num uint32
+	fd          int32
+	major       uint32
+	minor       uint32
+}
+
+const (
+	ILLEGAL_FD           = -1
```
Comment (AkihiroSuda) [inline libcontainer/vtpm/vtpm.go]: `IllegalFD` would be more Go-ish

File: libcontainer/vtpm/vtpm.go [hunk 4]
```diff
@@ -0,0 +1,893 @@
...
+func vtpmx_ioctl(cmd, msg uintptr) error {
+	vtpmx, err := os.Open("/dev/vtpmx")
+	if err != nil {
+		logrus.Warnf("Could not open /dev/vtpmx: %v", err)
+		return err
+	}
+	defer vtpmx.Close()
+
+	if err := ioctl(uintptr(vtpmx.Fd()), cmd, msg); err != nil {
+		return fmt.Errorf("VTPM: vtpmx ioctl failed: %v", err)
```
Comment (AkihiroSuda) [inline libcontainer/vtpm/vtpm.go]: `errors.Wrap`

File: vendor/github.com/opencontainers/runtime-spec/specs-go/config.go (hunk 1 above)
Comment (AkihiroSuda) [reply vendor/github.com/opencontainers/runtime-spec/specs-go/config.go]: What's current status of https://github.com/opencontainers/runtime-spec/pull/920 ?

cc @vbatts

**Comment by AkihiroSuda** [comment]
Seems we need to wait to see how https://github.com/opencontainers/runtime-spec/pull/920 will settle

**Comment by mfranczy** [comment]
@stefanberger are you still working on this PR?

**Comment by kolyshkin** [comment]
Since https://github.com/opencontainers/runtime-spec/pull/920 is closed, we should close this one as well.

**Pull request closed without merging**